core/src/test/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyImplTests.java

```diff
@@ -249,29 +250,34 @@ public void testBuilderWithDefaultRolePrefix() {
 			.implies("B")
 			.role("B")
 			.implies("C", "D")
+			.authority("C")
+			.implies("E", "F", "B")
 			.build();
-		List<GrantedAuthority> flatAuthorities = AuthorityUtils.createAuthorityList("ROLE_A");
-		List<GrantedAuthority> allAuthorities = AuthorityUtils.createAuthorityList("ROLE_A", "ROLE_B", "ROLE_C",
-				"ROLE_D");
 
-		assertThat(roleHierarchyImpl).isNotNull();
-		assertThat(roleHierarchyImpl.getReachableGrantedAuthorities(flatAuthorities))
-			.containsExactlyInAnyOrderElementsOf(allAuthorities);
+		assertHierarchy(roleHierarchyImpl).givesToAuthorities("ROLE_A")
+			.theseAuthorities("ROLE_A", "ROLE_B", "ROLE_C", "ROLE_D");
+
+		assertHierarchy(roleHierarchyImpl).givesToAuthorities("C")
+			.theseAuthorities("C", "ROLE_B", "ROLE_C", "ROLE_D", "ROLE_E", "ROLE_F");
```

Comment:
As of now, using `.authority("C").implies("E", "F", "B")`, means that `C` will have `ROLE_E`, `ROLE_F` and `ROLE_B`. What if I want to say that authority `user:write` implies in `user:read`, `user:block`, for example?

I'm thinking that when using `authority(...)` the role prefix should be ignored for the implied authorities too. It should be possible to do `.authority("user:write").implies("user:read", "user:delete")`. What do you think?

Reply:
My mindset when writing this was that 
- the authority is a "label" that is provided externally provided (via authentication).
- the role is a "label" that is internal to use to actually give access to things.

So from your example assuming the 'user:...' are externally provided authorities I see this as a possible way of doing this:

```
.authority("user:read")  .implies("USER_READ") 
.authority("user:write") .implies("USER_WRITE")
.authority("user:delete").implies("USER_DELETE")

.authority("TeamXYZMembers").implies("USER_READ", "USER_WRITE")

.role("USER_WRITE")      .implies("USER_READ", "USER_DELETE")
```

This would make all members of "Team XYZ" have the roles USER_READ, USER_WRITE and USER_DELETE.

Primary question: Does this reasoning make sense? It is correct?
If this is correct; Perhaps it should be documented as such?
(if you want I'll put up a separate merge request with such documentation)

The way I normally structure this is that in an application I define roles being groups of users with the same set of tasks like "Administrator", "Manager", "User", etc. Those are then assigned to the user in an external IGA/IAM system (often based on department/job title/...).

Then via the authentication of the specific user; the application receives one or more of these "external roles" as the authorities via the authentication.

From there my application has something like this
```
.authority("User")         .implies("USER_READ") 
.authority("Manager")      .implies("USER_WRITE")
.authority("Administrator").implies("USER_DELETE", "SHOW_DASHBOARD")

.role("USER_WRITE")       .implies("USER_READ")
.role("USER_DELETE")      .implies("USER_WRITE")
```

As said; I you want I'm willing to make a first draft on documenting this.

Reply:
If you want to be able to imply an authority then something like this is needed:
``` .authority("user:write").impliesAuthority("user:read", "user:delete")```

A separate function is needed because there is no way to make any distinction between `authority` or `role` because they are both just a string.

Also
- I currently disagree with this approach given how I look at the `authority` (external) and `role` (internal).
- It is not needed to get the desired effect as shown in the example I posted yesterday.